datafusion/physical-expr/src/expressions/binary.rs

```diff
@@ -1213,6 +1214,13 @@ pub fn binary(
 ) -> Result<Arc<dyn PhysicalExpr>> {
     let lhs_type = &lhs.data_type(input_schema)?;
     let rhs_type = &rhs.data_type(input_schema)?;
+    if (is_utf8_or_large_utf8(lhs_type) && is_timestamp(rhs_type))
```

Comment:
Thanks @Weijun-H  -- this does indeed fix the bug as stated. 

 I was trying to figure out how this code even executes (and was not caught earlier)

It seems like maybe this could be related to

https://github.com/apache/arrow-datafusion/blob/d8925bebc1f5382b044ce80f78bc465a9145e425/datafusion/expr/src/type_coercion/binary.rs#L80-L85

🤔 

Maybe it is time to figure out a better coercion story

Reply:
+1 sometime ago we agreed to that logical planner will be a better place for coercions and coercion checks